### PR TITLE
feat: Add `codersdk.NullTime`, change workspace build deadline

### LINKED ISSUE #2015: Codersdk: make workspace_build.deadline a pointer
# What

Change this

```
type WorkspaceBuild struct {
    ...
    Deadline time.Time `json:"deadline"`
    ...
```

To this:

```
type WorkspaceBuild struct {
    ...
    Deadline *time.Time `json:"deadline"`
    ...
```

# Why

The FE currently needs to check for the zero-value of `time.Time` for 'falsey' values.
Making this a pointer would allow passing `null` and avoid this need.

### PR DESCRIPTION
**Pull request by mafredri**
I took a slightly different approach in this PR than what originally proposed in #2015, it's closer to my proposal in a later comment.

Currently we embedd `sql.NullTime` for _one reason_, it gives us database scanner/valuer and means we can pass nullable times to the DB by doing `t.NullTime`.

We can simplify this though, if it's not needed so that we have plain `codersdk.NullTime{Time time.Time; Valid boold}`.

Fixes #2015

### CODE REVIEW

**Comment by mafredri** [comment]
// ping @johnstcn, what do you think about this approach?

File: codersdk/workspacebuilds.go [hunk 1]
```diff
@@ -50,7 +50,7 @@ type WorkspaceBuild struct {
 	InitiatorID        uuid.UUID           `json:"initiator_id"`
 	InitiatorUsername  string              `json:"initiator_name"`
 	Job                ProvisionerJob      `json:"job"`
-	Deadline           time.Time           `json:"deadline"`
+	Deadline           NullTime            `json:"deadline,omitempty"`
```
Comment (kylecarbs) [inline codersdk/workspacebuilds.go:53]: Based on the comment in the attached issue, it's unclear what the use-case for knowing the time is null is that a pointer wouldn't solve.

Is this related to the TypeScript generation?

**Review by kylecarbs — CHANGES_REQUESTED** [review]

File: codersdk/workspacebuilds.go (hunk 1 above)
Comment (mafredri) [reply codersdk/workspacebuilds.go:53]: I think this is a cleaner approach from a usability perspective in Go code. That is to say, the benefit is not in the frontend talking to the backend, it's using these values in the backend.

IMO it'd be nice for us to consistently use `NullTime` over `*time.Time` in the API. Things like not being able to take a pointer of a function call `&time.Now()` are not an issue when you can write `codersdk.NewNullTime(time.Now(), true)`. Likewise when using the type, there's no need for the `if t != nil` check.

There are work-arounds for making pointer use a bit more convenient, like using `coderd/util/ptr`, but it's an extra step to using them vs accessing the concrete type directly.

One last point in favor of `NullTime`, it's safer to use. There's no risk of doing a nil pointer dereference.

Comment (johnstcn) [reply codersdk/workspacebuilds.go:53]: FWIW working with pointers makes me sad; I much prefer the `NullXXX` approach.
Unless we start going down the lines of having a generic `Optional[T]` type?

**Review by johnstcn — APPROVED** [review]

File: codersdk/workspacebuilds.go (hunk 1 above)
Comment (kylecarbs) [reply codersdk/workspacebuilds.go:53]: I'm hesitant to stray from such strong language primitives. Do we imagine having a `NullXXX` for every type that could be nullable? This seems odd to me because embedded structs would likely be pointers with `omitempty` rather than `Null<StructName>`. Are we promoting this *only* for built-in types?

Comment (mafredri) [reply codersdk/workspacebuilds.go:53]: I understand why you may be hesitant, but I don't think it will be a problem. If we strive to make the zero value useful in most cases, the need for nullable values will be pretty low. We can also look at `codersdk` (today), there we only have a couple of primitives (int, string) and time.

```
      7 *int64
      5 *time.Time
      3 *string
      2 *uuid.UUID
```

The `uuid` package already has `uuid.NullUUID`, so it would make sense to use that one instead. In fact, the occurrence of `uuid.NullUUID` is much larger in our codebase than that of `*uuid.UUID` (30 vs 3), yet we have both.

I also foresee that the need for a nullable struct will be very rare (we have none in `codersdk`, currently), and in those cases it would likely be a custom type for which we can simply implement a custom json marshaler/unmarshaler. It's unfortunate Go never got around to adding `omitzero`/`omitempty` to the `json` package, because many times that would solve the use-case for using pointer-structs (i.e. often the reason is not to detect omission, it's to create a specific type of output).

**Review by jsjoeio — APPROVED** [review]
✅ FE
 
I only see an update to the `deadline` property being made optional now. I imagine this shouldn't have any consequences assuming we have tests for code using that interface. 
 
 If there is something else you want reviewed by FE, ping me!

**Review by kylecarbs — APPROVED** [review]

**Review by jsjoeio — APPROVED** [review]
Thanks for going the extra mile to double-check any side effects on the FE! 👏🏼

File: site/src/components/WorkspaceScheduleButton/WorkspaceScheduleButton.tsx [hunk 2]
```diff
@@ -28,11 +28,7 @@ dayjs.extend(relativeTime)
 dayjs.extend(timezone)
 
 export const shouldDisplayPlusMinus = (workspace: Workspace): boolean => {
-  if (!isWorkspaceOn(workspace)) {
-    return false
-  }
-  const deadline = dayjs(workspace.latest_build.deadline).utc()
-  return deadline.year() > 1
+  return isWorkspaceOn(workspace) && !!workspace.latest_build.deadline
```
[the remark this answers is not in the document]
Comment (jsjoeio) [reply site/src/components/WorkspaceScheduleButton/WorkspaceScheduleButton.tsx:31]: > Should we make this change here already, if it's not decided yet?

Up to you! I approved and prefixed with `nit` to convey that but missed the mark 🤣 

> I did a quick comparison and !! seems more commonly used than Boolean for the time being.

Sounds good to me then 👍🏼 

> but at the same time it feels a bit weird to sometimes use ! and other times Boolean. 😄

Oh I totally agree - I'm all about consistency!

[the remark this answers is not in the document]
Comment (mafredri) [reply site/src/components/WorkspaceScheduleButton/WorkspaceScheduleButton.tsx]: Haha right on! 😄 I made the change, rooting for your initiative. 👍🏻

[the remark this answers is not in the document]
Comment (jsjoeio) [reply site/src/components/WorkspaceScheduleButton/WorkspaceScheduleButton.tsx]: Woo! Team `Boolean` making strides 😎
